### PR TITLE
LibWeb: Make ExceptionOr work with non-JS::Value types 

### LINKED ISSUE #6075: LibWeb: ExceptionOr doesn't work with values that JS::Value doesn't accept
I'm currently rewriting the node mutation algorithms to be spec compliant and this is a common return type: `ExceptionOr<NonnullRefPtr<Node>>`

However, this doesn't quite work:
```
In file included from Userland/Libraries/LibWeb/Bindings/NodePrototype.cpp:19:
../../Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h: In instantiation of 'Ret Web::Bindings::throw_dom_exception_if_needed(auto:19&&, auto:20&&, F&&) [with F = Web::Bindings::NodePrototype::append_child(JS::VM&, JS::GlobalObject&)::<lambda()>; T = Web::DOM::ExceptionOr<AK::NonnullRefPtr<Web::DOM::Node> >; Ret = JS::Value; auto:19 = JS::VM&; auto:20 = JS::GlobalObject&]':
Userland/Libraries/LibWeb/Bindings/NodePrototype.cpp:350:108:   required from here
../../Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h:62:33: error: could not convert '(& result)->Web::DOM::ExceptionOr<AK::NonnullRefPtr<Web::DOM::Node> >::value()' from 'AK::NonnullRefPtr<Web::DOM::Node>' to 'JS::Value'
   62 |             return result.value();
      |                                 ^
      |                                 |
      |                                 AK::NonnullRefPtr<Web::DOM::Node>
../../Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h: In instantiation of 'Ret Web::Bindings::throw_dom_exception_if_needed(auto:19&&, auto:20&&, F&&) [with F = Web::Bindings::NodePrototype::insert_before(JS::VM&, JS::GlobalObject&)::<lambda()>; T = Web::DOM::ExceptionOr<AK::NonnullRefPtr<Web::DOM::Node> >; Ret = JS::Value; auto:19 = JS::VM&; auto:20 = JS::GlobalObject&]':
Userland/Libraries/LibWeb/Bindings/NodePrototype.cpp:395:113:   required from here
../../Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h:62:33: error: could not convert '(& result)->Web::DOM::ExceptionOr<AK::NonnullRefPtr<Web::DOM::Node> >::value()' from 'AK::NonnullRefPtr<Web::DOM::Node>' to 'JS::Value'
../../Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h: In instantiation of 'Ret Web::Bindings::throw_dom_exception_if_needed(auto:19&&, auto:20&&, F&&) [with F = Web::Bindings::NodePrototype::remove_child(JS::VM&, JS::GlobalObject&)::<lambda()>; T = Web::DOM::ExceptionOr<AK::NonnullRefPtr<Web::DOM::Node> >; Ret = JS::Value; auto:19 = JS::VM&; auto:20 = JS::GlobalObject&]':
Userland/Libraries/LibWeb/Bindings/NodePrototype.cpp:427:107:   required from here
../../Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h:62:33: error: could not convert '(& result)->Web::DOM::ExceptionOr<AK::NonnullRefPtr<Web::DOM::Node> >::value()' from 'AK::NonnullRefPtr<Web::DOM::Node>' to 'JS::Value'
```

### PR DESCRIPTION
**Pull request by alimpfard**
Fixes #6075.

cc @Lubrsi: Please double-check the exceptions I returned; I used the comments, but one case was just "throw an exception", not sure if I did the right thing.

### CODE REVIEW

File: Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h [hunk 1]
```diff
@@ -49,30 +50,61 @@ ALWAYS_INLINE bool throw_dom_exception(JS::VM& vm, JS::GlobalObject& global_obje
...
 
 template<typename T>
-bool should_return_empty(T&& value)
+bool should_return_empty(const Optional<T>& value)
 {
     if constexpr (IsSame<JS::Value, T>)
-        return value.is_empty();
-    return false;
+        return !value.has_value() || value.value().is_empty();
+    return !value.has_value();
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h:107]: I'm not sure if this is 100% right, we're gonna be returning empty values in these cases:
- The operation threw an exception
- The operation returned void

I'm not sure if we're supposed to be returning `undefined` for the second case, or an empty value?

(cc @linusg)

Comment (linusg) [reply Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h:107]: > The operation threw an exception

Empty value is fine.

> The operation returned void

Needs to be undefined I think, as `undefined foo();` in IDL maps to `void foo();` in C++.

Comment (alimpfard) [reply Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h:107]: Alright, I'll make it do that then : )

**Review by Lubrsi — APPROVED** [review]
LGTM
> one case was just "throw an exception", not sure if I did the right thing.

It's correct, I just wanted to return the exception given by `ensure_pre_insertion_validity`. The reason I put it like that is because `ensure_pre_insertion_validity` returns different types of exceptions. (namely, `NotFoundError` and `HierarchyRequestError`)

**Comment by linusg** [comment]
```
The following tests FAILED:
	  4 - Shell-backgrounding (Timeout)
```

I think I made you test shell backgrounding and that accidentally ended up in this PR... :grimacing:

**Comment by alimpfard** [comment]
:facepalm:

**Comment by awesomekling** [comment]
Oh very rad!

**Comment by awesomekling** [comment]
Now we can get really serious about throwing the correct exceptions in every little corner case :smile:

**Pull request merged**